README.md
# Hammer

Hammer is a theme development tool for the [CleanSlate](http://cleanslate.wvu.edu/ "CleanSlate") CMS template rendering engine. Hammer was created to help you iterate quickly on themes and template markup.

Hammer is **NOT** for content creation.

## Install & Usage

Please install via [Hammer-VM](https://github.com/wvuweb/hammer-vm)

## Interested? Need Help? Read more..

Visit the wiki for more information:

[Hammer Wiki](https://github.com/wvuweb/hammer/wiki)

### Development Dependencies

These dependencies are only needed if you are developing against Hammer directly

* Ruby 2.4
* Bundler Gem `gem install bundler`

### Mock Data File

Hammer uses a YAML file to 'mock' content data structures needed for cleanslate templates.

[`mock_data.yml`](https://github.com/wvuweb/hammer/wiki/Mock-Data#example-of-a-mock_datayml-file)

### Mock Data Documentation

```YAML
hammer_nav:                                     # hammer navigation controls
  disabled: Boolean                             # disabled                    -- accepts true | false, default: false
                                                # hides the navigation bar to allow testing for accessibility

edit_mode: Boolean                               # edit mode                   -- accepts true | false
                                                # edit mode allows you to hide and show content based on context
                                                # <r:edit_mode_only>
                                                #   <p>This only shows in the editor mode of CleanSlate</p>
                                                # </r:edit_mode_only>

shared_themes:                                  # shared_themes are defined by
                                                # <r:partial name="file/path" theme="theme-name"/>
  "theme-name":                                 # theme-name                  -- accepts quote theme name string
    "file/path": theme                          # file/path                   -- accepts quoted file path string
  "theme-name-2":                               # theme                       -- accepts string
    "file/path": theme                                                 

editable_region:                                # editable_region are defined by
  region: String                                # <r:editable_region name="region" />
  region2: String                               # region                      -- accepts a string or | pipe character followed by HTML
                                                # example:
                                                # editable_region:
                                                #   region: |
                                                #     <p>Hello World!</p>

site:                                           # site is defined by any attributes your site would hold in CleanSlate
  id: 26                                        # id <r:site:id />            -- accepts an integer
  name: String                                  # name <r:site:name />        -- accepts a string
  domain: example.wvu.edu                       # domain <r:site:domain />    -- accepts a string
  data:                                         # data is defined by any custom data your page would hold in CleanSlate
    data_key: data_value                        # <r:site:data name="data_key" />
    data_key_2: data_value_2                    # data_key is the name of the custom data in page properties which is defined
                                                # in your templates front matter configuration
                                                # data_value is any value you wish to expose
root: true                                      # root tag is used to set the context for loops
                                                # <r:root />                  -- accepts true | false

page:                                           # page is defined by any attribute a page would hold in CleanSlate
  id: 1                                         # <r:page:id />               -- accepts an integer
  name: String                                  # <r:page:name />             -- accepts a string
  slug: string-dashed-notation                  # <r:page:slug />             -- accepts a string in dash notation
  meta_description: String                      # <r:page:meta_description /> -- accepts a string
  title: String                                 # <r:page:title />            -- accepts a string
  alternate_name: String                        # <r:page:alternate_name />   -- accpets a string
  depth: 1                                      # <r:page:depth />            -- accepts an integer
  updated_at: Jan 1st 2015 2:30PM               # <r:page:updated_at />       -- accepts a date string
  created_at: two weeks ago                     # <r:page:created_at />       -- accepts a date string
  published_at: now                             # <r:page:published_at />     -- accepts a date string
  tags:                                         # <r:page:children:each tags="tag1,tag2" tag_op="=" />
    - String                                    # Tags accepts an Array of Strings
    - String
  data:                                         # data is defined by any custom data your page would hold in CleanSlate
    data_key: data_value                        # <r:data name="data_key" />
    data_key_2: data_value_2                    # data_key is the name of the custom data in page properties which is defined
                                                # in your templates front matter configuration
                                                # data_value is any value you wish to expose
  content:                                      # content attribute is for supporting loops in which a page
    region_name: String                         # may be dynamically building content from child pages or <r:get_page />
                                                # <r:page:if_has_content_for name="region_name" />
                                                # <r:page:unless_has_content_for name="region_name" />
pages:                                          # pages allow you to mock addtional pages to be used in page loops
  - id: 2                                       # such as <r:children /> <r:ancestors /> <r:decendants /> <r:siblings />
    name: String                                # each page set is seperated by a dash to create an Array of pages
    ...                                         # pages can contain all the data that the "page" key above can
  - id: 3
    ...
  - id: 4
    ...

if_page_depth_eq: 1                             # <r:if_page_depth_eq page_depth="1"/> test for page depth equal to value
if_page_depth_gt: 1                             # <r:if_page_depth_gt page_depth="1"/> test for page depth greater than value

site_menu: String                               # site menu allows you to override the html generated if you have a pages block
                                                # <r:site_menu />             -- accepts a string or | pipe character followed by HTML
                                                # example:
                                                # site_menu: |
                                                #   <ul>
                                                #     <li class="active"><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #   </ul>

sub_menu: String                                # sub_menu allows you to override the auto generated html created from the pages block
                                                # or the generic menu that gets output if you do not have a pages block.
                                                # <r:sub_menu />              -- accepts a string or | pipe character followed by HTML
                                                # example:
                                                # sub_menu: |
                                                #   <ul>
                                                #     <li class="active"><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #   </ul>

ancestor_menu: String                           # ancestor_menu allows you to override the auto generated html created from the pages block
                                                # or the generic menu that gets output if you do not have a pages block.
                                                # <r:ancestor_menu/>          -- accepts a string or | pipe character followed by HTML
                                                # example:
                                                # sub_menu: |
                                                #   <ul>
                                                #     <li class="active"><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #     <li><a href="#">Page</a></li>
                                                #   </ul>

breadcrumbs: String                             # breadcrumbs takes a string of HTML that outputs breadcrumb structure
                                                # breadcrumbs: |
                                                #   <ul class="wvu-breadcrumbs__crumbs">
                                                #     <li><a href="#">Home</a></li>
                                                #     <li class="active">Page</li>
                                                #   </ul>

files:                                          # files contains all the data for any of the <r:files|file /> tags
  - filename: String                            # files is an Array of files seperated by a dash
    name: String
    title: String
    alt_text: String
    download_url: String
    image_url: String
  - ...

blogs:                                          # blogs contains all the data for the <r:blog:{method} /> tags
  - id: Integer                                 # blogs contains an Array of blogs each blog set is seperated by a dash
    name: String                                # <r:blog:name />             -- accepts a string
                                                # blog articles much like pages are a repeating group of content for each article
    articles:                                   # each article is seperated by a dash  
      - article:                                # <r:blog:articles /> loops can access the following data in a <r:each />
        name: String                            # <r:article:name />          -- accepts a string
        title: String                           # <r:article:title />         -- accepts a string
        created_by:                             # <r:article:author_full_name />  -- accepts two other objects  
          first_name: String                    # <r:article:author_first_name /> -- accepts a string
          last_name: String                     # <r:article:author_last_name />  -- acccpts a string
        content:                                # content attribute is for supporting loops in which a blog article
          region_name: String                   # may be dynamically built using content regions from the article
                                                # <r:article:content name="region_name" />  
        tags:                                   # <r:articles tags="tag1,tag2" tags_op="any|none">
          - String                              # tags takes an array of String values
          - String
        published_at: 2 days ago                # <r:article:published_at />  -- accepts a date string
      - article:
        ...
      - article:
        ...
  - id: 2
    name: String
    ...
```

## Developer notes
Changelog uses [Github Changelog Generator](https://github.com/github-changelog-generator/github-changelog-generator)

[Generate a token](https://github.com/settings/tokens/new?description=GitHub%20Changelog%20Generator%20token) use in the command below in the project root.

`github_changelog_generator wvuweb/hammer --token {insert your token here} --future-release {release_version}`


CHANGELOG.md
# Change Log

## [v1.1.5](https://github.com/wvuweb/hammer/tree/v1.1.5) (2019-04-16)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.1.4...v1.1.5)

**Implemented enhancements:**

- Support tags on pages [\#86](https://github.com/wvuweb/hammer/issues/86)

## [v1.1.4](https://github.com/wvuweb/hammer/tree/v1.1.4) (2019-04-15)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.1.3...v1.1.4)

**Fixed bugs:**

- r:root incorrectly looks for a page context [\#83](https://github.com/wvuweb/hammer/issues/83)
- r:if and r:content\_for cause duplicate output [\#82](https://github.com/wvuweb/hammer/issues/82)
- Calling a missing yaml partial causes white screen of death without a helpful error message. [\#81](https://github.com/wvuweb/hammer/issues/81)

## [v1.1.3](https://github.com/wvuweb/hammer/tree/v1.1.3) (2019-04-08)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.1.2...v1.1.3)

**Implemented enhancements:**

- Include better error backtracing when there is a partial error [\#80](https://github.com/wvuweb/hammer/issues/80)
- Better syntax surrounding shared\_theme partials [\#79](https://github.com/wvuweb/hammer/issues/79)
- Add better depreciation notices [\#78](https://github.com/wvuweb/hammer/issues/78)
- Add version to hammer toolbar [\#75](https://github.com/wvuweb/hammer/issues/75)

## [v1.1.2](https://github.com/wvuweb/hammer/tree/v1.1.2) (2019-03-28)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.1.1...v1.1.2)

**Implemented enhancements:**

- Tag feature needs documented [\#74](https://github.com/wvuweb/hammer/issues/74)

**Fixed bugs:**

- Blog article content missing key does not display parent key [\#73](https://github.com/wvuweb/hammer/issues/73)

## [v1.1.1](https://github.com/wvuweb/hammer/tree/v1.1.1) (2019-03-28)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.1.0...v1.1.1)

**Fixed bugs:**

- Blog tags if comma delimited attribute contains a space [\#72](https://github.com/wvuweb/hammer/issues/72)
- Hammer not reporting correct version. [\#71](https://github.com/wvuweb/hammer/issues/71)

## [v1.1.0](https://github.com/wvuweb/hammer/tree/v1.1.0) (2019-03-27)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.14...v1.1.0)

**Implemented enhancements:**

- Add the ability to use tag attribute on blogs [\#70](https://github.com/wvuweb/hammer/issues/70)

**Fixed bugs:**

- \<r:article:id /\> outputs duplicate ID's in a loop [\#69](https://github.com/wvuweb/hammer/issues/69)

## [v1.0.14](https://github.com/wvuweb/hammer/tree/v1.0.14) (2019-03-22)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.13...v1.0.14)

**Implemented enhancements:**

- Add environment for loading only development dependencies [\#68](https://github.com/wvuweb/hammer/issues/68)
- When hammer is run in daemon mode, log errors [\#67](https://github.com/wvuweb/hammer/issues/67)
- Hammer version check reports fatal error with git describe in dev mode [\#66](https://github.com/wvuweb/hammer/issues/66)

## [v1.0.13](https://github.com/wvuweb/hammer/tree/v1.0.13) (2019-03-20)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.12...v1.0.13)

**Implemented enhancements:**

- Add a flag in mock\_data to show/hide the Hammer breadcrumb toolbar [\#65](https://github.com/wvuweb/hammer/issues/65)
- Change ARIA Label on Hammer \<nav\> element. [\#64](https://github.com/wvuweb/hammer/issues/64)
- Give a more helpful error message when the blog array in mock data isn't formatted properly [\#59](https://github.com/wvuweb/hammer/issues/59)
- Add support for attributes to r:files [\#36](https://github.com/wvuweb/hammer/issues/36)

## [v1.0.12](https://github.com/wvuweb/hammer/tree/v1.0.12) (2019-03-19)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.11...v1.0.12)

**Implemented enhancements:**

- Changelog line should always display [\#62](https://github.com/wvuweb/hammer/issues/62)

**Fixed bugs:**

- Changelog is incorrectly located [\#61](https://github.com/wvuweb/hammer/issues/61)

## [v1.0.11](https://github.com/wvuweb/hammer/tree/v1.0.11) (2019-03-19)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.10...v1.0.11)

## [v1.0.10](https://github.com/wvuweb/hammer/tree/v1.0.10) (2019-03-19)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.9...v1.0.10)

## [v1.0.9](https://github.com/wvuweb/hammer/tree/v1.0.9) (2019-03-19)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.8...v1.0.9)

**Fixed bugs:**

- Hammer incorrectly checking and displaying latest tagged version [\#60](https://github.com/wvuweb/hammer/issues/60)

## [v1.0.8](https://github.com/wvuweb/hammer/tree/v1.0.8) (2019-03-19)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.7...v1.0.8)

## [v1.0.7](https://github.com/wvuweb/hammer/tree/v1.0.7) (2019-03-15)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.6...v1.0.7)

## [v1.0.6](https://github.com/wvuweb/hammer/tree/v1.0.6) (2019-03-15)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.5...v1.0.6)

## [v1.0.5](https://github.com/wvuweb/hammer/tree/v1.0.5) (2019-03-14)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.4...v1.0.5)

## [v1.0.4](https://github.com/wvuweb/hammer/tree/v1.0.4) (2019-03-14)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.3...v1.0.4)

## [v1.0.3](https://github.com/wvuweb/hammer/tree/v1.0.3) (2019-03-14)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.2...v1.0.3)

## [v1.0.2](https://github.com/wvuweb/hammer/tree/v1.0.2) (2019-03-14)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.1...v1.0.2)

## [v1.0.1](https://github.com/wvuweb/hammer/tree/v1.0.1) (2019-03-14)
[Full Changelog](https://github.com/wvuweb/hammer/compare/v1.0.0...v1.0.1)

## [v1.0.0](https://github.com/wvuweb/hammer/tree/v1.0.0) (2019-03-06)
**Implemented enhancements:**


hammer/services/tags/blog.rb
require 'chronic'
require_relative "../tag_container.rb"

module Tags

  # Blog Tag Module
  class Blog < TagContainer

    def initialize
      @blog = nil
    end

    include ActionView::Helpers::TagHelper
    include ActionView::Context

    BOOLEAN_ATTRIBUTES = %w(disabled readonly multiple checked autobuffer
                            autoplay controls loop selected hidden scoped async
                            defer reversed ismap seamless muted required
                            autofocus novalidate formnovalidate open pubdate
                            itemscope allowfullscreen default inert sortable
                            truespeed typemustmatch).to_set
    BOOLEAN_ATTRIBUTES.merge(BOOLEAN_ATTRIBUTES.map{|attribute| attribute.to_sym})
    PRE_CONTENT_STRINGS = {
      textarea: "\n"
    }

    tag 'blog' do |tag|
      tag.locals.blog ||= load_blog(tag)
      errors = tag.locals.blog_errors
      content = []
      errors.each do |error|
        content << error
      end
      if tag.locals.blog
        content << tag.expand
      end
      content.join("")
    end

    tag 'article' do |tag|
      content = []
      if tag.locals.article_errors
        tag.locals.article_errors.each do |error|
          content << error
        end
      end
      if tag.locals.article
        content << tag.expand
      end
      content.join("")
    end

    tag 'article:id' do |tag|
      if tag.locals.article['id']
        tag.locals.article['id']
      else
        Hammer.key_missing "id", {parent_key: "article"}
      end
    end

    tag 'article:name' do |tag|
      if tag.locals.article['name']
        tag.locals.article['name']
      else
        Hammer.key_missing "name", {parent_key: "article"}
      end
    end

    tag 'article:title' do |tag|
      if tag.locals.article['title']
        tag.locals.article['title']
      else
        Hammer.key_missing "title", {parent_key: "article"}
      end
    end

    tag 'article:path' do |tag|
      if tag.locals.article['url']
        tag.locals.article['url']
      else
        tag.render 'page:url', tag.attr
      end
    end

    tag 'article:content' do |tag|
      # tag.render 'page:content', tag.attr
      region = tag.attr['name']
      if tag.locals.article[:content]
        if tag.locals.article[:content][region]
          tag.locals.article[:content][region]
        else
          if region == 'article-body'
            output = []
            output << (Hammer.key_missing region, {parent_key: 'article:content', warning: true, message: 'auto-generated paragraphs have been inserted below'})
            Faker::Lorem.paragraphs(rand(1..3), true).each do |paragraph|
              output << "<p>#{paragraph}</p>"
            end
            output.join("")
          else
            Hammer.key_missing region, {parent_key: 'article:content'}
          end
        end
      else
        Hammer.key_missing 'content', {parent_key: 'article'}
      end
    end

    tag 'article:tags' do |tag|
      # tag.locals.article.label_list.join(',')
      if tag.locals.article['tags'].kind_of?(Array)
        tag.locals.article['tags'].join(',')
      elsif
        Hammer.error 'Article tags should stored as an array in mock_data.yml'
      end
    end

    tag 'article:published_at' do |tag|
      if tag.locals.article[:published_at]
        tag.locals.article[:published_at]
      else
        "#{(0..10).to_a.sample} days ago"
      end
    end

    tag 'article:author_first_name' do |tag|
      if tag.locals.article[:created_by][:first_name]
        tag.locals.article[:created_by][:first_name]
      else
        Hammer.key_missing "first_name", {parent_key: "article"}
      end
    end

    tag 'article:author_last_name' do |tag|
      if tag.locals.article[:created_by][:last_name]
        tag.locals.article[:created_by][:last_name]
      else
        Hammer.key_missing "last_name", {parent_key: "article"}
      end
    end

    tag 'article:author_full_name' do |tag|
      if tag.locals.article[:created_by]
        if tag.locals.article[:created_by][:first_name] || tag.locals.article[:created_by][:last_name]
          [tag.locals.article[:created_by][:first_name], tag.locals.article[:created_by][:last_name]].join(" ")
        else
          content = []
          content << (Hammer.key_missing "first_name or last_name", {parent_key: "article:created_by", comment: true, warning: true})
          content << [Faker::Name.first_name, Faker::Name.last_name].join(" ")
          content.join("")
        end
      else
        content = []
        content << (Hammer.key_missing "created_by", {parent_key: "article", comment: true})
        content << [Faker::Name.first_name, Faker::Name.last_name].join(" ")
        content.join("")
      end
    end

    tag 'articles' do |tag|
      load_blog(tag)
      tag.locals.articles = filter_articles(tag, tag.locals.blog[:articles])
      tag.expand
    end

    tag 'articles:each' do |tag|
      loop_over tag, tag.locals.articles
    end

    tag 'articles:count' do |tag|
      count_items tag, tag.locals.articles
    end

    tag 'articles:if_articles' do |tag|
      # cnt = tag.locals.articles.try(:all).try(:count)
      # tag.expand if cnt > 0

      tag.expand if tag.locals.articles.count > 0
    end

    tag 'articles:if_no_articles' do |tag|
      # cnt = tag.locals.articles.try(:all).try(:count)
      # tag.expand if cnt.nil? or cnt == 0
      if tag.locals.articles.count.nil? || tag.locals.articles.count == 0
        tag.expand
      end
    end

    tag 'articles:pagination' do |tag|
      tag.expand
      # Hammer.error "articles:pagination tag is not implemented yet"
    end

    tag 'articles:pagination:previous_url' do |tag|
      url_for_page(tag, :previous)
    end

    tag 'articles:pagination:next_url' do |tag|
      url_for_page(tag, :next)
    end

    tag 'articles:pagination:if_first_page' do |tag|
      return unless tag.locals.article_pagination.present?
      tag.expand if tag.locals.article_pagination[:previous].nil?
    end

    tag 'articles:pagination:if_last_page' do |tag|
      return unless tag.locals.article_pagination.present?
      tag.expand if tag.locals.article_pagination[:next].nil?
    end

    tag 'archive' do |tag|
      tag.expand
    end

    tag 'archive:monthly' do |tag|
      options = gather_options(
        tag,
        {
          date_format: '%B %Y',
          ul_class: 'cs-blog-archive--monthly',
          include_count: true,
          count_class: 'cs-blog-archive__count'
        }
      )

      ActionView::Base.new.content_tag :ul, class: options[:ul_class] do
        if tag.locals.blog[:archive] && tag.locals.blog[:archive][:monthly] && tag.locals.blog[:archive][:monthly].count > 0
          data = tag.locals.blog[:archive][:monthly]
        else
          date_to = Date.parse(Chronic.parse('today').strftime("%Y-%m-%d").to_s)
          date_from = Date.parse(Chronic.parse('4 months ago').strftime("%Y-%m-%d").to_s)

          date_range = date_from..date_to
          date_months = date_range.map {|d| Date.new(d.year, d.month, 1) }.uniq

          data = [
                  {"item"=>{"date"=>"#{date_months[0]}", "count"=>20, "url"=>"#"}},
                  {"item"=>{"date"=>"#{date_months[1]}", "count"=>14, "url"=>"#"}},
                  {"item"=>{"date"=>"#{date_months[2]}", "count"=>32, "url"=>"#"}},
                  {"item"=>{"date"=>"#{date_months[3]}", "count"=>22, "url"=>"#"}},
                  {"item"=>{"date"=>"#{date_months[4]}", "count"=>18, "url"=>"#"}}
                ].reverse
        end

        data.map do |a|
          date = parse_date(a['item']['date'])
          ActionView::Base.new.content_tag :li do
            content = "#{date.strftime(options[:date_format])}"
            content += ' ' + ActionView::Base.new.content_tag(:span, a['item']['count'], class: options[:count_class]) if options[:include_count].to_s.to_b
            ActionView::Base.new.content_tag :a, content.html_safe, href: a['item']['url']
          end
        end.join("\n").html_safe
      end
    end

    def self.parse_date(str)
      # We will try parsing the date string with Chronic first. If that doesn't work, we'll attempt to
      # parse the string with DateTime.
      Time.zone = "Eastern Time (US & Canada)"
      Chronic.time_class = Time.zone
      Chronic.parse(str) || DateTime.parse(str) rescue nil
    end


    class << self
      def gather_options(tag, defaults = {})
        options = defaults.symbolize_keys
        attrs = (%w(ul_id ul_class).map(&:to_sym) + defaults.keys.map(&:to_sym)).uniq

        attrs.each do |opt|
          options[opt] = tag.attr[opt.to_s] if tag.attr.has_key?(opt.to_s)
        end

        options
      end

      def load_blog(tag)
        tag.locals.blog = []
        tag.locals.blog_errors = []

        if tag.globals.context.data['page']
          tag.locals.page = tag.globals.context.data['page']
        else
          tag.locals.blog_errors << (Hammer.key_missing "page")
        end

        blogs = tag.globals.context.data['blogs'] || tag.globals.context.data['blog']

        if tag.globals.context.data['blog']
          tag.locals.blog_errors << (Hammer.error "In future releases <code>blog:</code> key will be renamed <code>blogs:</code> and be of type Array ", {comment: true, depreciation: true})
        end

        if blogs
          if blogs.kind_of?(Array)
            # Match current page id to blog id
            if blogs.select{|w| w['id'].to_s == (tag.locals.page['id'].to_s) }.first
              tag.locals.blog = blogs.select{|w| w['id'].to_s == (tag.locals.page['id'].to_s) }.first
            else
              tag.locals.blog_errors << (Hammer.error "Could not find blog with id: #{tag.locals.page['id'].to_s} in mock_data.yml")
            end
          else
            tag.locals.blog_errors << (Hammer.error "In future releases <code>blog:</code> key will be renamed <code>blogs:</code> and be of type Array", {comment: true, depreciation: true})
            tag.locals.blog = blogs
          end
        else
          tag.locals.blog_errors << (Hammer.key_missing "blogs")
        end
      end

      def filter_articles(tag, target)
        if tag.attributes && tag.attributes['tags'] && !tag.attributes['tags'].empty?
          # split tags on comma and remove whitespace from each item
          tags = tag.attributes['tags'].split(',').collect{|x| x.strip }
          tagged_items = []

          if tag.attributes['tags_op'] && tag.attributes['tags_op'] == 'none'
            reverse = true
          end

          # get all articles that include any of the tags
          target.each do |art|
            # artile tag in mock_data.yml may not exist if
            # just return empty array if that is the case

            article_tags = art[:tags] || []
            # remote whitespace that may exist in mock data
            article_tags = article_tags.collect{|x| x.strip }
            if article_tags
              # if the tags_ops is none then you want all articles without tags
              if reverse
                if (tags & article_tags).empty?
                  tagged_items << art
                end
              else
                unless (tags & article_tags).empty?
                  tagged_items << art
                end
              end
            end
          end
          target = tagged_items
        end
        # limit should be last
        if tag.attributes && tag.attributes['limit']
          limit = tag.attributes['limit'].to_i - 1
          target = target[0..limit]
        end
        target
      end

      def count_items(tag, target)
        tag.locals.blog[:articles].count
      end

      def loop_over(tag, target)
        items = filter_articles(tag, target)

        output = []

        items.each_with_index do |item, index|
          tag.locals.index = index + 1
          tag.locals.page = item
          tag.locals.article = item
          output << tag.expand
        end

        output.flatten.join('')
      end

      def url_for_page(tag, key)
        '#'
      end

    end

  end
end
